libs/switches.liq

```diff
@@ -0,0 +1,166 @@
+           pick()
+           !picked_index == index
+        end
+      else
+        false
+      end
+    end
+
+
+    (cond, s)
+  end
+
+  s = switch(override=override, replay_metadata=replay_metadata, track_sensitive=true,
+             transition_length=transition_length, transitions=transitions,
+             list.mapi(add_condition, sources))
+
+  fallback(id=id, track_sensitive=true, list.add(s, sources))
+end
+
+# At the beginning of every track, select a random ready child.
+# @category Source / Track Processing
+# @param ~id Force the value of the source ID.
+# @param ~override Metadata field which, if present and containing a float, overrides the `transition_length` parameter.
+# @param ~replay_metadata Replay the last metadata of a child when switching to it in the middle of a track.
+# @param ~transition_length Maximun transition duration.
+# @param ~transitions Transition functions, padded with `fun (x,y) -> y` functions.
+# @param ~weights Weights of the children (padded with 1), defining for each child the probability that it is selected.
+def random(~id="", ~override="liq_transition_length", ~replay_metadata=true,
+           ~transition_length=5., ~transitions=[],
+           ~weights=[], sources) =
+  weights = list.map(to_int_getter, weights)
+  default_weigth = fun () -> 1
+  next_index = ref -1
+
+  def pick() =
+    def available_weighted_sources(cur, source) =
+      let (index, current_weight, indexes) = cur
+      weight = list.nth(default=default_weigth, weights, index)
+
+      let (current_weight, indexes) =
+        if source.is_ready(source) then
+          weight = weight()
+          indexes = list.add((current_weight, current_weight+weight, index), indexes)
+          (current_weight + weight, indexes)
+        else
+          (current_weight, indexes)
+        end
+    
+      (index+1, current_weight, indexes)
+    end
+
+    let (_, total_weight, weighted_indexes)  =
+      list.fold(available_weighted_sources, (0, 0, []), sources)
+
+    picked_weight =
+      if total_weight > 0 then
+        random.int(min=0, max=total_weight)
+      else
+        0
+      end
+      
+    def pick_source(picked_index, el) =
+      let (lower_bound, upper_bound, index) = el
+
+      if lower_bound <= picked_weight and picked_weight < upper_bound then
+        index
+      else
+         picked_index
+      end
+    end
+ 
+    next_index := list.fold(pick_source, -1, weighted_indexes)
+ end
+
+  def add_condition(index, s) =
+    def f(_) =
+      next_index := -1
+    end 
+
+    s = on_track(f, s)
+
+    def cond() =
+      if !next_index == -1 then pick() end
+      !next_index == index  
+    end
+    
+    (cond, s)
+  end
+
+  s = switch(override=override, replay_metadata=replay_metadata, track_sensitive=true,
+             transition_length=transition_length, transitions=transitions,
+             list.mapi(add_condition, sources))
+
+  fallback(id=id, track_sensitive=true, list.add(s, sources))
+end
```

Comment:
The similarity between `rotate` and `random` makes me wonder if `switch` is really the right "basic" function. Wouldn't it be simpler to take as common abstraction a function which takes as argument

- a list of pairs `('a, source)` (where 'a is some information about the source (the weights, etc.)
- a "picker" function of type `[('a, source)] -> source` which returns the next source to be played

and returns a source. All the above functions could then be implemented from this one.